### PR TITLE
Guaranteed delivery for message store producer - ESBJAVA-3852

### PR DESCRIPTION
**Pull request by prabathariyaratna**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+			                     Integer.parseInt((String) parametersMap.get(MessageProcessorConstants.MAX_DELIVER_ATTEMPTS));
+		}
+
+		if (parametersMap.get(MessageProcessorConstants.RETRY_INTERVAL) != null) {
+			retryInterval =
+			                Integer.parseInt((String) parametersMap.get(MessageProcessorConstants.RETRY_INTERVAL));
+		}
+
+		faultSeq = (String) parametersMap.get(FailoverForwardingProcessorConstants.FAULT_SEQUENCE);
+
```
Comment (jagathsisira) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Avoid unwanted blank lines

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+		// Setting the interval value.
+		interval = Long.parseLong((String) parametersMap.get(MessageProcessorConstants.INTERVAL));
+
+		/*
+		 * Make sure to set the isInitialized flag to TRUE in order to avoid
+		 * re-initialization.
+		 */
+		initialized = true;
+	}
+
```
Comment (jagathsisira) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Avoid unwanted blank lines

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+	 *
+	 * @param messageContext
+	 *            synapse {@link MessageContext} to be sent
+	 */
+	public void dispatch(MessageContext messageContext) {
+		if (log.isDebugEnabled()) {
+			log.debug("Sending the message to client with message processor [" +
+			          messageProcessor.getName() + "]");
+		}
+
```
Comment (jagathsisira) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Avoid unwanted blank lines

File: modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java [hunk 4]
```diff
@@ -85,14 +85,47 @@ public boolean storeMessage(MessageContext synCtx) {
...
+
+            try {
+
+                if (session.getTransacted()) {
+                    session.rollback();
+                }
+
+            } catch (JMSException ex) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Fail to rollback message [" + synCtx.getMessageID() + "] from the message store " +
```
Comment (jagathsisira) [inline modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java]: Better to preserve exact exception in the log

File: modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java [hunk 5]
```diff
@@ -85,14 +85,47 @@ public boolean storeMessage(MessageContext synCtx) {
...
+
+            try {
+
+                if (session.getTransacted()) {
+                    session.rollback();
+                }
+
+            } catch (JMSException e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Fail to rollback message [" + synCtx.getMessageID()+"] from the message store " +
```
Comment (jagathsisira) [inline modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java]: Better to preserve exact exception in the log

File: modules/core/src/main/java/org/apache/synapse/mediators/store/MessageStoreMediator.java [hunk 6]
```diff
@@ -49,11 +51,32 @@
      */
     private String  onStoreSequence;
 
+    /**
+     * Status of the guaranteed delivery
+     */
+    private boolean isGuaranteedDeliveryEnable = false;
```
Comment (nirmal070125) [inline modules/core/src/main/java/org/apache/synapse/mediators/store/MessageStoreMediator.java]: /s/isGuaranteedDeliveryEnable/isGuaranteedDeliveryEnabled

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+	private int maxDeliverAttempts = -1;
+	private int attemptCount = 0;
+
+    private boolean isThrottling = true;
+
+    /**
+     * Throttling-interval is the forwarding interval when cron scheduling is enabled.
+     */
+    private long throttlingInterval = -1;
+
```
Comment (nirmal070125) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Class seems to be not formatted.

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java (hunk 1 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java (hunk 2 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java (hunk 3 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java (hunk 4 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java (hunk 5 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/store/impl/jms/JmsProducer.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/mediators/store/MessageStoreMediator.java (hunk 6 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/mediators/store/MessageStoreMediator.java]: Fixed in https://github.com/wso2/wso2-synapse/pull/373

**Comment by prabathariyaratna** [comment]
Formatting issues fix in https://github.com/wso2/wso2-synapse/pull/373

File: modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java (hunk 7 above)
Comment (prabathariyaratna) [inline modules/core/src/main/java/org/apache/synapse/message/processor/impl/failover/FailoverForwardingService.java]: Fixed in ps://github.com/wso2/wso2-synapse/pull/373

**Pull request merged**